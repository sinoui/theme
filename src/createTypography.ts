import deepMerge from 'deepmerge';
import { TypographyConstants, ThemeTypography, FontWeight } from './types';

const DEFAULT_FONT_FAMILY = `-apple-system, BlinkMacSystemFont, 'Segoe UI', 'PingFang SC', 'Hiragino Sans GB',
'Microsoft YaHei', Roboto, 'Noto Sans SC', 'Helvetica Neue', Helvetica, Arial, sans-serif, 'Apple Color Emoji',
'Segoe UI Emoji', 'Segoe UI Symbol'`;

const allCapsStyle: React.CSSProperties = {
  textTransform: 'uppercase',
};

const getStandardTypography = (
  fontWeightLight: FontWeight,
  fontWeightRegular: FontWeight,
  fontWeightMedium: FontWeight,
  pxToRem: (size: number) => string,
) => {
  const buildStyle = (
    fontWeight: FontWeight,
    size: number,
    lineHeight: number,
    letterSpacing: number,
    style?: React.CSSProperties,
  ): React.CSSProperties => ({
    fontWeight,
    fontSize: pxToRem(size),
    lineHeight,
    letterSpacing: `${Math.round((letterSpacing / size) * 1e5) / 1e5}em`,
    ...style,
  });

  const standard = {
    h1: buildStyle(fontWeightLight, 96, 1.167, -1.5),
    h2: buildStyle(fontWeightLight, 60, 1.2, -0.5),
    h3: buildStyle(fontWeightRegular, 48, 1.167, 0),
    h4: buildStyle(fontWeightRegular, 34, 1.235, 0.25),
    h5: buildStyle(fontWeightRegular, 24, 1.334, 0),
    h6: buildStyle(fontWeightMedium, 20, 1.6, 0.15),
    subtitle1: buildStyle(fontWeightRegular, 16, 1.75, 0.15),
    subtitle2: buildStyle(fontWeightMedium, 14, 1.57, 0.1),
    body1: buildStyle(fontWeightRegular, 16, 1.5, 0.15),
    body2: buildStyle(fontWeightRegular, 14, 1.43, 0.15),
    button: buildStyle(fontWeightMedium, 15, 1.75, 0.4, allCapsStyle),
    caption: buildStyle(fontWeightRegular, 12, 1.66, 0.4),
    overline: buildStyle(fontWeightRegular, 12, 2.66, 1, allCapsStyle),
  };

  return standard;
};

/**
 * 创建文本排版
 *
 * @param constants 文本排版配置
 */
export default function createTypography(
  constants: TypographyConstants = {},
): ThemeTypography {
  const {
    fontFamily = DEFAULT_FONT_FAMILY,
    fontWeightLight = 300,
    fontWeightRegular = 400,
    fontWeightMedium = 500,
    fontWeightBold = 700,
    fontSize = 14,
    htmlFontSize = 16,
    pxToRem = (size: number) => `${((size / 14) * fontSize) / htmlFontSize}rem`,
  } = constants;

  const theme = getStandardTypography(
    fontWeightLight,
    fontWeightRegular,
    fontWeightMedium,
    pxToRem,
  );

  return deepMerge(
    {
      ...theme,
      fontFamily,
      fontWeightRegular,
      fontWeightLight,
      fontWeightMedium,
      fontWeightBold,
      fontSize,
      htmlFontSize,
      pxToRem,
    },
    constants,
  );
}
