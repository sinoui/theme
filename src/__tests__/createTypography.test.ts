import createTypography from '../createTypography';

it('创建默认的字体排版', () => {
  const typography = createTypography();

  expect(typography).toMatchSnapshot();
});

it('通过fontSize控制字体大小', () => {
  const typography = createTypography({
    fontSize: 12,
  });

  expect(typography.body2.fontSize).toBe('0.75rem');
  expect(typography.fontSize).toBe(12);
});

it('设置 htmlFontSize', () => {
  const typography = createTypography({
    htmlFontSize: 10,
  });

  expect(typography.body1.fontSize).toBe('1.6rem');
  expect(typography.htmlFontSize).toBe(10);
});

it('设置 fontFamily', () => {
  const typography = createTypography({
    fontFamily: 'Noto Sans SC',
  });

  expect(typography.fontFamily).toBe('Noto Sans SC');
});

it('设置字体类型样式', () => {
  const typography = createTypography({
    h1: {
      color: 'red',
    },
  });

  expect(typography.h1).toEqual({
    fontSize: '6rem',
    fontWeight: 300,
    letterSpacing: '-0.01562em',
    lineHeight: 1.167,
    color: 'red',
  });
});

it('使用 pxToRem 函数', () => {
  const { pxToRem } = createTypography();

  expect(pxToRem(16)).toBe('1rem');
});
