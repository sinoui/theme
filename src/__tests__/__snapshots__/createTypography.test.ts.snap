// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`创建默认的文字排版 1`] = `
Object {
  "body1": Object {
    "fontSize": "1rem",
    "fontWeight": 400,
    "letterSpacing": "0.00938em",
    "lineHeight": 1.5,
  },
  "body2": Object {
    "fontSize": "0.875rem",
    "fontWeight": 400,
    "letterSpacing": "0.01071em",
    "lineHeight": 1.43,
  },
  "button": Object {
    "fontSize": "0.875rem",
    "fontWeight": 500,
    "letterSpacing": "0.02857em",
    "lineHeight": 1.75,
    "textTransform": "uppercase",
  },
  "caption": Object {
    "fontSize": "0.75rem",
    "fontWeight": 400,
    "letterSpacing": "0.03333em",
    "lineHeight": 1.66,
  },
  "fontFamily": "-apple-system, BlinkMacSystemFont, 'Segoe UI', 'PingFang SC', 'Hiragino Sans GB',
'Microsoft YaHei', Roboto, 'Noto Sans SC', 'Helvetica Neue', Helvetica, Arial, sans-serif, 'Apple Color Emoji',
'Segoe UI Emoji', 'Segoe UI Symbol'",
  "fontSize": 14,
  "fontWeightBold": 700,
  "fontWeightLight": 300,
  "fontWeightMedium": 500,
  "fontWeightRegular": 400,
  "h1": Object {
    "fontSize": "6rem",
    "fontWeight": 300,
    "letterSpacing": "-0.01562em",
    "lineHeight": 1.167,
  },
  "h2": Object {
    "fontSize": "3.75rem",
    "fontWeight": 300,
    "letterSpacing": "-0.00833em",
    "lineHeight": 1.2,
  },
  "h3": Object {
    "fontSize": "3rem",
    "fontWeight": 400,
    "letterSpacing": "0em",
    "lineHeight": 1.167,
  },
  "h4": Object {
    "fontSize": "2.125rem",
    "fontWeight": 400,
    "letterSpacing": "0.00735em",
    "lineHeight": 1.235,
  },
  "h5": Object {
    "fontSize": "1.5rem",
    "fontWeight": 400,
    "letterSpacing": "0em",
    "lineHeight": 1.334,
  },
  "h6": Object {
    "fontSize": "1.25rem",
    "fontWeight": 500,
    "letterSpacing": "0.0075em",
    "lineHeight": 1.6,
  },
  "htmlFontSize": 16,
  "overline": Object {
    "fontSize": "0.75rem",
    "fontWeight": 400,
    "letterSpacing": "0.08333em",
    "lineHeight": 2.66,
    "textTransform": "uppercase",
  },
  "pxToRem": [Function],
  "subtitle1": Object {
    "fontSize": "1rem",
    "fontWeight": 400,
    "letterSpacing": "0.00938em",
    "lineHeight": 1.75,
  },
  "subtitle2": Object {
    "fontSize": "0.875rem",
    "fontWeight": 500,
    "letterSpacing": "0.00714em",
    "lineHeight": 1.57,
  },
}
`;
